has/haselements.go

```diff
@@ -0,0 +1,51 @@
+package has
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/corbym/gocrest"
+)
+
+// Checks wether each element of the array/slice matches each exptectation passed to ElementsWith
+// Panics if the actual is not an array/slice
+func ElementsWith(expects ...*gocrest.Matcher) *gocrest.Matcher {
+	match := new(gocrest.Matcher)
+	match.Matches = func(actual interface{}) bool {
+
+		actualValue := reflect.ValueOf(actual)
+		switch actualValue.Kind() {
+		case reflect.Array, reflect.Slice:
+			reason, actual := elementsMatch(expects, actualValue)
+
+			match.Describe = reason
+			match.Actual = actual
+			return reason == "" && actual == ""
```

Comment:
I am not entirely sure we should be mixing the concerns of matcher descriptions and matchers matching here.

Using the `reason` and `actual` of the passed in matchers to determine if this matcher matched seems a bit fragile (some homemade custom matchers may not append actuals at all, for example), and it confused me at least as to what was happening here.

So a better way would be to build the matcher's `Describe` first using all the matchers available, then work out whether or not the matcher passes. You can see a similar example in `is.AllOf` matcher:

```
func AllOf(allMatchers ...*gocrest.Matcher) *gocrest.Matcher {
	matcher := new(gocrest.Matcher)
	matcher.Describe = fmt.Sprintf("all of (%s)", describe("and", allMatchers)) // builds the description of the matcher
	matcher.Matches = matchAll(allMatchers, matcher) // returns true if all the matchers match
	return matcher
}
```

The 'actual' of this matcher should be appended to the `ElementsWith` matcher, in a similar way to `AllOf`: 

```
func matchAll(allMatchers []*gocrest.Matcher, allOf *gocrest.Matcher) func(actual interface{}) bool {
	return func(actual interface{}) bool {
		matches := true
		for x := 0; x < len(allMatchers); x++ {
			if !allMatchers[x].Matches(actual) {
				matches = false
			}
			allOf.AppendActual(allMatchers[x].Actual)
		}
		return matches
	}
```

`allOf.AppendActual` adds all the matchers actuals to the 'but was:' output string. There is an argument that the `allOf.AppendActual` should not be done in this function either (as it is not the same concern) - so I am lecturing myself here too lol.

Reply:
fixed `.Describe`.
Both methods now call `AppendActual` for each of their matchers.